### DISCUSSION TITLE
stumpy.match bug with parameters: Sliding mean and Sliding standard deviation

### DISCUSSION THREAD

**Discussion by brunopinos31** [TDAmeritrade/stumpy#581] [Help: Coding & Implementations] [answered]
I have a bug when i try to use the stumpy.match function with parameters: Sliding mean and Sliding standard deviation.

![image](https://user-images.githubusercontent.com/92527271/161106075-5853222b-5073-4e57-8cda-e75d1abe171b.png)

**Comment by seanlaw** [comment, marked as answer]
@brunopinos31 Thank you for reporting this and welcome to the STUMPY community. I will take a look

**Reply by seanlaw** [reply]
@brunopinos31 Indeed, there was a minor bug. I have committed a fix for this and will push a new version in the coming days. Thanks again for bringing it to our attention!

**Reply by seanlaw** [reply]
@brunopinos31 I have released v1.11.1 that fixes this.

**Reply by brunopinos31** [reply]
Perfect it works! Thank you seanlaw !

**Reply by seanlaw** [reply]
May I ask what the use case for adding those paramters is? Typically, they are needed if you are planning to compare many, many different queries, `Q`, against the same time series, `T`, and this may save a bit of time especially if your time series is very long. Essentially, you are only saving on the time needed to recompute the mean/stddev for `T`. However, if you're only querying a handful of times then it may not really be worth the decreased code readability.

**Reply by brunopinos31** [reply]
I calculate the number of non trivial match for each subsequence of a time series for a fixed radius. So I compare many, many different queries, Q, against the same time series, T.

But maybe there is a way to improve the speed of my algo.

Here my code:
![image](https://user-images.githubusercontent.com/92527271/161264338-27510ad9-047e-4837-847c-1fa2baf53467.png)

**Reply by seanlaw** [reply]
This seems like a matrix profile anti-pattern to me since you are essentially computing ALL pairwise distances in your time series using `stumpy.match`. This can be very slow when your time series is over 50K data points in length and very slow using `stumpy.match`. Additionally, this is essentially what the matrix profile is doing (computed using `stumpy.stump`) but it only returns the one-nearest-neighbor (it still computes the distance to all other subsequences outside of the exclusion zone). How long is your time series?

The usual pattern is to first compute the matrix profile and then only use `stumpy.match` on the subsequences in your time series where the matrix profile distance is less than `4.0` (i.e., your `max_distance`). This is useful since the matrix profile gives you the distance to your one-nearest-neighbor and so if that one-nearest-neighbor is a distance of `4.1` units away then that also implies that the distance to ALL other subsequences MUST be larger than `4.1`. In other words, there must be zero matches for this subsequence when you use `stumpy.match` with `max_distance=4.0`. Instead, you use the information stored in the matrix profile to help you filter down to only the relevant subsequences (i.e., only those subsequences that have at least one-nearest-neighbor less than or equal to `max_distance`) that you can then further interrogate using `stumpy.match`. Here is some pseudocode:

```
# untested code
mp = stumpy.stump(T, m)
counter = []
max_distance = 4.0
for i in range(mp.shape[0]):
    if mp[i] <= max_distance
        Q = T[i:i+m]
        matches = stumpy.match(Q, T, mean, stddev, max_distance=max_distance)
        counter.append(matches)
```

Hopefully, I am not misunderstanding your point/goal.

**Reply by brunopinos31** [reply]
You understood well ;)

Oh thank you good idea. Is there a way to speed up the algo by telling stumpy.match that it doesn't need to calculate the distance with subsequences whose nearest neighbor distance is greater than 4? Other than breaking down the time series into several time series.

Why building a kind of profile matrix that counts the number of subsequences at a distance less than r for each subsequence can't be done at the same speed as the classic matrix profile?

**Reply by seanlaw** [reply]
>  Is there a way to speed up the algo by telling stumpy.match that it doesn't need to calculate the distance with subsequences whose nearest neighbor distance is greater than 4? Other than breaking down the time series into several time series.

No, I am not aware of any other simple way. You may be able to ignore large sections of your time series after you've computed the matrix profile but you'll need to do some careful bookkeeping. How long is your time series? How many subsequences have a one-nearest-neighbor with a distance less than 4.0?

> Why building a kind of profile matrix that counts the number of subsequences at a distance less than r for each subsequence can't be done at the same speed as the classic matrix profile?

This is because `core.mass` leverages Fast Fourier Transforms and only reduce the computational time down to `O(n x n x log n)` while `stumpy.stump` is based on the STOMP algorithm which can compute the matrix profile in `O(n x n)` time. So, `stumpy.stump` should, theoretically, be faster by a factor of `log n` (where `n` is the total length of your time series). For more information, I recommend reading/comparing the first two matrix profile papers:

* [Matrix Profile I](https://www.cs.ucr.edu/~eamonn/PID4481997_extend_Matrix%20Profile_I.pdf)
* [Matrix Profile II](https://www.cs.ucr.edu/~eamonn/STOMP_GPU_final_submission_camera_ready.pdf)

**Reply by brunopinos31** [reply]
Ok thank you for all these informations! My algo will be use on different time series for different usecases.

stumpy.motif will have 'variable length' implementation ?

**Reply by seanlaw** [reply]
> stumpy.match and stumpy.motif will have 'variable length' implementation ?

I don't understand the meaning of your question. What "variable length" quantity are you trying to compute? Can you please clarify what you mean by your question?

**Reply by brunopinos31** [reply]
I mean the length of the motif.

For stumpy.motif:
input: pan matrix profile computed with stimp
output: motifs of different length

**Reply by seanlaw** [reply]
So, unless you are computing the pan matrix profile with `percentage=1.0` (e.g., `stumpy.stimp(T, min_m=3, max_m=10, percentage=1.0)`), I don't think it's a good idea to be using the pan matrix profile to find motifs as any value of `percentage < 1.0` is only an approximation of the exact matrix profile. As a consequence, any motifs obtained from an approximate pan matrix profile would also be approximate. However, if you are are computing the pan matrix profile with `percentage=1.0` then you might as well just do:

```
mps = []
for m in range(3, 11):
    mps.append(stumpy.stump(T, m))
```

and then apply `stumpy.motifs` to each matrix profile.

**Reply by brunopinos31** [reply]
Ok thx Seanlaw. 
I'm looking for an implementation of [VALMOD](https://helios2.mi.parisdescartes.fr/~themisp/publications/sigmod18-valmod.pdf) in python.
I don't know if it exists^^

**Reply by seanlaw** [reply]
You're correct, we do not have a VALMOD implementation. There is also this other related paper on [MAD](https://link.springer.com/article/10.1007/s10618-020-00685-w). I haven't read them yet but it may be relevent

**Reply by brunopinos31** [reply]
I may try to do something for valmod I'll let you know. Thank you for all !
